conwayste/src/ui/layer.rs

```diff
@@ -260,25 +264,42 @@ impl Layering {
         let node = self.widget_tree.get_mut(&inserted_node_id).unwrap();
         node.data_mut().set_id(inserted_node_id.clone());
 
+        // Report the error but keep going
+        if self.removed_node_ids.contains(&inserted_node_id) {
+            error!("NodeId {:?} found in HashSet during widget insertion. Possible reusage! Please report the bug!", inserted_node_id);
+        }
+
         Ok(inserted_node_id)
     }
 
     /// Removes a widget belonging to the layering. Will drop all child nodes if the target is a
     /// container-based widget.
     ///
-    /// # Error
+    /// # Errors
+    ///
     /// A WidgetNotFound error can be returned if a widget with the `widget_id` does not exist
     /// in the internal list of widgets.
     // Implemented API for future use. TODO: Remove comment once function is used
     #[allow(unused)]
     pub fn remove_widget(&mut self, id: NodeId) -> UIResult<()> {
-        if !self.check_for_entry(&id) {
+        if !self.widget_exists(&id) {
             return Err(Box::new(UIError::WidgetNotFound {
                 reason: format!("{:?} not found in layer during removal", id).to_owned(),
             }));
         }
 
-        self.widget_tree.remove_node(id, RemoveBehavior::DropChildren);
+        // clone is okay because it is required
+        self.widget_tree.remove_node(id.clone(), RemoveBehavior::DropChildren).or_else(|e| {
+            return Err(Box::new(UIError::InvalidAction {
+                // clone is okay for error reporting
+                reason: format!("NodeIDError occurred during removal of {:?}: {:?}", id.clone(), e)
+            }));
+        });
+
+        // clone is okay because the HashSet is intended to keep track of all removed widgets
+        if !self.removed_node_ids.insert(id.clone()) {
+            error!("NodeId {:?} found in HashSet during widget removal. Possible reusage! Please report the bug!", id);
```

Comment:
If it's already in the set, I say just ignore this. It already got logged when the widget was added.